docs/mergerepo.1
.TH "mergerepo" "1" "2008 Oct 21" "Seth Vidal" ""

.SH "NAME"
mergerepo \- Merge multiple repositories together

.SH "SYNOPSIS"
\fBmergerepo\fP \-\-repo repo1 \-\-repo repo2
.PP

.SH "DESCRIPTION"
\fBmergerepo\fP is a program that allows you merge multiple repositories
into a single repository while referring to the remote location for all
packages.

.SH "OPTIONS"
.IP "\fB\-r \-\-repo\fP <url>"
Url to a repository to be merged.

.IP "\fB\-o \-\-outputdir <directory>\fP"
Path where merged repository metadata should be written to. If not specified
repository metadata will be written to `pwd`/merged_repo/.

.IP "\fB\-d \-\-database\fP"
Generate sqlite databases of the merged repository metadata. This is now the default.

.IP "\fB\-a \-\-archlist\fP"
Specify a comma-separated list of architectures to use. Defaults to ALL.

.IP "\fB\-\-nogroups\fP"
Do not merge/include groups metadata in the repository.

.IP "\fB\-\-noupdateinfo\fP"
Do not merge/include updateinfo metadata in the repository.

.IP "\fB\-\-no-database\fP"
Do not generate sqlite databases in the repository.

.IP "\fB\-\-compress-type <compress-type>\fP"
Specify which compression type to use: compat (default), xz (may not be available), gz, bz2.

.IP "\fB\-\-version\fP"
Output version.

.IP "\fB\-h \-\-help\fP"
Show help menu.

.SH "EXAMPLES"
.PP
$ \fBmergerepo\fP \-\-repo=http://myurl.org/repo1 \-\-repo=http://myurl.org/repo2 \-d \-o /tmp/mymergedrepo

.PP 
.SH "SEE ALSO"
.I createrepo (8)

.PP 
.SH "AUTHORS"
.nf 
Seth Vidal <skvidal@fedoraproject.org>
.fi 

.PP 
.SH "BUGS"
Any bugs which are found should be emailed to the mailing list:
rpm-metadata@lists.baseurl.org or filed as tickets at:
http://createrepo.baseurl.org/

.fi


ChangeLog
2014-01-28  Zdenek Pavlas <zpavlas@redhat.com>
	* createrepo --update: ignore cached rpm when checksum_type
	  has changed.

2014-01-16  Zdenek Pavlas <zpavlas@redhat.com>

	* catch IOErrors when loading pkglist. BZ 1044997
	* Add missing int() on mtime
	* Fix file/dir completions for names containing spaces or tabs

2013-11-28  Zdenek Pavlas <zpavlas@redhat.com>

	* Removing the workers=1 default breaks API users, put it back.
	* Make sure the "packages" attribute is always correct.
	* Clean up tempdir on failures.
	* Close lock file before unlink()
	* mark as 0.10.1

2013-10-18  Zdenek Pavlas <zpavlas@redhat.org>

	* mark as 0.10

2011-01-26  Seth Vidal <skvidal@fedoraproject.org>

	* createrepo.spec, createrepo/__init__.py: mark as 0.9.9

2011-01-26  Seth Vidal <skvidal@fedoraproject.org>

	Merge branch 'master' of
	ssh://createrepo.baseurl.org/srv/projects/createrepo/git/createrepo
	* 'master' of
	ssh://createrepo.baseurl.org/srv/projects/createrepo/git/createrepo:
	Override timestamp check on repos. for mergerepo (like repodiff).
	Add createrepo --workers (non)completion.   Add modifyrepo option
	completion.

2011-01-21  James Antill <james@and.org>

	* createrepo/merge.py: Override timestamp check on repos. for
	mergerepo (like repodiff).

2011-01-03  Seth Vidal <skvidal@fedoraproject.org>

	* createrepo/__init__.py: make sure when we want to look for rpms we
	say .rpm not rpm b/c with the latter we catch .drpm files, too. :(

2010-11-02  Ville Skyttä <ville.skytta@iki.fi>

	* createrepo.bash: Add createrepo --workers (non)completion.

2010-11-02  Ville Skyttä <ville.skytta@iki.fi>

	* createrepo.bash: Add modifyrepo option completion.

2010-10-08  Seth Vidal <skvidal@fedoraproject.org>

	* createrepo.spec, createrepo/__init__.py: - add yum 3.2.29 requirement b/c of the small change I needed to
	repoMDObject.py -  set it to use /usr/share/createrepo/worker.py

2010-10-07  Seth Vidal <skvidal@fedoraproject.org>

	* createrepo/__init__.py: remove libxml2 import from __init__.py :)


2010-10-07  Seth Vidal <skvidal@fedoraproject.org>

	* createrepo/__init__.py: make createrepo use the repomd/repodata
	mechanism from yum for making a repomd.xml which simplifies the code
	dramatically since we don't have to mess with xml in here.

2010-10-07  Seth Vidal <skvidal@fedoraproject.org>

	* modifyrepo.py: fix up the usage output for modifyrepo

2010-09-10  Seth Vidal <skvidal@fedoraproject.org>

	* createrepo/__init__.py, worker.py: - make sure we handle remote_url pkgs correctly until we get the
	worker hooked up to handle them - if there are no pkgs to handle,
	don't launch workers with nothing to do. - give better output from
	the workers and have them obey -v/-q - everyone loves callbacks!

2010-09-09  Seth Vidal <skvidal@fedoraproject.org>

	* Makefile, createrepo/__init__.py, createrepo/utils.py,
	createrepo/yumbased.py, genpkgmetadata.py, worker.py: create a
	worker script for createrepo so createrepo can fork off N processes
	to handle the md gathering from pkgs. This should speed up results
	on systems which have been cpubound on the createrepo process.  If
	you're io bound it won't help you at all, and MAY make it worse.
	many misc issues to iron out here - not the least of which is the
	callback output and gathering stdout/stderr from the workers

2010-08-20  Seth Vidal <skvidal@fedoraproject.org>

	* createrepo/__init__.py: handle broken locking on nfs target dirs
	better if database is true.  - sqlite dbs don't like being made on
	locations without locking available. - if we know we're going to be
	creating dbs then we should attempt to lock before doing   anything
	else and bail out nicely if we can't get an exclusive lock

2010-08-19  Seth Vidal <skvidal@fedoraproject.org>

	* createrepo.spec: require yum 3.2.28 due to the imports in
	modifyrepo

2010-08-19  Seth Vidal <skvidal@fedoraproject.org>

	* docs/modifyrepo.1: document --mdtype option

2010-08-19  Seth Vidal <skvidal@fedoraproject.org>

	* modifyrepo.py: - add option parsing for --mdtype - use the yum repomd objects to
	read/write the repomd.xml - add mdtype option to RepoMetadata.add()
	method

2010-06-11  Seth Vidal <skvidal@fedoraproject.org>

	Merge branch 'master' of
	ssh://createrepo.baseurl.org/srv/projects/createrepo/git/createrepo
	* 'master' of
	ssh://createrepo.baseurl.org/srv/projects/createrepo/git/createrepo:
	Don't use /usr/bin/env ... it's evil   --database is now the default
	for mergerepo too, have --no-database in completions instead.

2010-06-11  Seth Vidal <skvidal@fedoraproject.org>

	* createrepo/__init__.py: - add option to createrepo config to collapse libc.so.6 requires -
	this will only work with yum 3.2.28 and beyond

2010-06-03  James Antill <james@and.org>

	* modifyrepo.py: Don't use /usr/bin/env ... it's evil

2010-06-02  Ville Skyttä <ville.skytta@iki.fi>

	* createrepo.bash: --database is now the default for mergerepo too,
	have --no-database in completions instead.

2010-06-01  Seth Vidal <skvidal@fedoraproject.org>

	* mergerepo.py: whoops - no-database shouldn't default to true!

2010-06-01  Seth Vidal <skvidal@fedoraproject.org>

	* mergerepo.py: add --no-database to mergrepo, too

2010-05-31  Ville Skyttä <ville.skytta@iki.fi>

	* createrepo.bash: --database is now the default, have --no-database
	in completions instead.

2010-05-28  Seth Vidal <skvidal@fedoraproject.org>

	* docs/createrepo.8: update the docs for --no-database

2010-05-28  Seth Vidal <skvidal@fedoraproject.org>

	* createrepo/__init__.py, genpkgmetadata.py: make -d/--database  the
	default  add --no-database in case someone somewhere needs to do
	that

2010-04-26  Seth Vidal <skvidal@fedoraproject.org>

	* createrepo/__init__.py: fixme comments about try/excepting the
	database creation calls due to a weird issue with locks not working
	on a nfs mount and createreepo tracing back with a TypeError of all
	things

2010-04-21  Seth Vidal <skvidal@fedoraproject.org>


mergerepo.py
#!/usr/bin/python -tt
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Library General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
# Copyright 2008  Red Hat, Inc - written by skvidal at fedoraproject.org

# merge repos from arbitrary repo urls

import sys
import createrepo.merge
from createrepo.utils import MDError
from optparse import OptionParser

#TODO:
# excludes?
# handle content/distro tags
# support revision?


def parse_args(args):
    """Parse our opts/args"""
    usage = """
    mergerepo: take 2 or more repositories and merge their metadata into a new repo

    mergerepo --repo=url --repo=url --outputdir=/some/path"""

    parser = OptionParser(version = "mergerepo 0.1", usage=usage)
    # query options
    parser.add_option("-r", "--repo", dest='repos', default=[], action="append",
                      help="repo url")
    parser.add_option("-a", "--archlist", default=[], action="append",
                      help="Defaults to all arches - otherwise specify arches")
    parser.add_option("-d", "--database", default=True, action="store_true")
    parser.add_option( "--no-database", default=False, action="store_true", dest="nodatabase")
    parser.add_option("-o", "--outputdir", default=None,
                      help="Location to create the repository")
    parser.add_option("", "--nogroups", default=False, action="store_true",
                      help="Do not merge group(comps) metadata")
    parser.add_option("", "--noupdateinfo", default=False, action="store_true",
                      help="Do not merge updateinfo metadata")
    parser.add_option("", "--xz", default=False, action="store_true",
                      help="Use xz for repodata compression")
                      
    (opts, argsleft) = parser.parse_args(args)

    if len(opts.repos) < 2:
        parser.print_usage()
        sys.exit(1)

    # sort out the comma-separated crap we somehow inherited.
    archlist = []
    for archs in opts.archlist:
        for arch in archs.split(','):
            archlist.append(arch)

    opts.archlist = archlist

    return opts

def main(args):
    """main"""
    opts = parse_args(args)
    rmbase = createrepo.merge.RepoMergeBase(opts.repos)
    if opts.archlist:
        rmbase.archlist = opts.archlist
    if opts.outputdir:
        rmbase.outputdir = opts.outputdir
    if opts.nodatabase:
        rmbase.mdconf.database = False
    if opts.nogroups:
        rmbase.groups = False
    if opts.noupdateinfo:
        rmbase.updateinfo = False
    if opts.xz:
        rmbase.mdconf.xz = True
    try:
        rmbase.merge_repos()
        rmbase.write_metadata()
    except MDError, e:
        print >> sys.stderr, "Could not merge repos: %s" % e
        sys.exit(1)
        
if __name__ == "__main__":
    main(sys.argv[1:])
